File: artifacts/api-server/src/routes/roi.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import { roiMetricsTable, ticketsTable } from "@workspace/db";
import { desc, eq } from "drizzle-orm";

const router: IRouter = Router();

router.get("/v1/roi/current", async (_req, res): Promise<void> => {
  const [latest] = await db.select().from(roiMetricsTable).orderBy(desc(roiMetricsTable.calculated_at)).limit(1);

  if (!latest) {
    const tickets = await db.select().from(ticketsTable);
    const autoResolved = tickets.filter(t => t.resolution_type === "auto").length;
    const total = tickets.length;
    const hoursSaved = autoResolved * (45 / 60);
    const costSaved = hoursSaved * 500;

    res.json({
      metric_id: "current",
      period_start: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10),
      period_end: new Date().toISOString().slice(0, 10),
      total_tickets: total,
      auto_resolved_count: autoResolved,
      manual_resolved_count: total - autoResolved,
      avg_auto_resolution_mins: 12.5,
      avg_manual_resolution_mins: 45.0,
      hours_saved: hoursSaved,
      cost_saved: costSaved,
      agent_hourly_cost: 500,
      calculated_at: new Date().toISOString(),
    });
    return;
  }

  res.json({
    metric_id: latest.metric_id,
    period_start: latest.period_start,
    period_end: latest.period_end,
    total_tickets: latest.total_tickets,
    auto_resolved_count: latest.auto_resolved_count,
    manual_resolved_count: latest.manual_resolved_count,
    avg_auto_resolution_mins: latest.avg_auto_resolution_mins ?? null,
    avg_manual_resolution_mins: latest.avg_manual_resolution_mins ?? null,
    hours_saved: latest.hours_saved,
    cost_saved: latest.cost_saved,
    agent_hourly_cost: latest.agent_hourly_cost,
    calculated_at: latest.calculated_at.toISOString(),
  });
});

router.get("/v1/roi/history", async (_req, res): Promise<void> => {
  const metrics = await db.select().from(roiMetricsTable).orderBy(desc(roiMetricsTable.calculated_at)).limit(12);

  if (metrics.length === 0) {
    const history = Array.from({ length: 6 }, (_, i) => {
      const date = new Date(Date.now() - (5 - i) * 30 * 24 * 60 * 60 * 1000);
      const autoResolved = Math.floor(Math.random() * 80) + 20;
      const hoursSaved = autoResolved * 0.75;
      return {
        metric_id: `mock-${i}`,
        period_start: new Date(date.getTime() - 30 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10),
        period_end: date.toISOString().slice(0, 10),
        total_tickets: autoResolved + Math.floor(Math.random() * 30),
        auto_resolved_count: autoResolved,
        manual_resolved_count: Math.floor(Math.random() * 30),
        avg_auto_resolution_mins: 12.5,
        avg_manual_resolution_mins: 45.0,
        hours_saved: hoursSaved,
        cost_saved: hoursSaved * 500,
        agent_hourly_cost: 500,
        calculated_at: date.toISOString(),
      };
    });
    res.json(history);
    return;
  }

  res.json(metrics.map(m => ({
    metric_id: m.metric_id,
    period_start: m.period_start,
    period_end: m.period_end,
    total_tickets: m.total_tickets,
    auto_resolved_count: m.auto_resolved_count,
    manual_resolved_count: m.manual_resolved_count,
    avg_auto_resolution_mins: m.avg_auto_resolution_mins ?? null,
    avg_manual_resolution_mins: m.avg_manual_resolution_mins ?? null,
    hours_saved: m.hours_saved,
    cost_saved: m.cost_saved,
    agent_hourly_cost: m.agent_hourly_cost,
    calculated_at: m.calculated_at.toISOString(),
  })));
});

router.patch("/v1/roi/settings", async (req, res): Promise<void> => {
  res.json({ success: true, message: "ROI settings updated" });
});

export default router;

File: artifacts/api-server/app/routers/roi.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, case

from app.database import get_db
from app.models.config import ROIMetric
from app.models.tickets import Ticket
from app.middleware.auth import get_current_agent

router = APIRouter()

DEFAULT_AGENT_HOURLY_COST = 25.0
DEFAULT_AVG_MANUAL_MINS = 45.0
DEFAULT_AVG_AUTO_MINS = 3.0


@router.get("/current")
async def current_roi(db: AsyncSession = Depends(get_db), agent=Depends(get_current_agent)):
    # Get latest metric or compute on-the-fly
    metric_result = await db.execute(
        select(ROIMetric).order_by(ROIMetric.calculated_at.desc()).limit(1)
    )
    metric = metric_result.scalar_one_or_none()

    # Live computation
    stats = await db.execute(
        select(
            func.count(Ticket.ticket_id).label("total"),
            func.sum(case((Ticket.status == "auto_resolved", 1), else_=0)).label("auto_resolved"),
        )
    )
    row = stats.one()
    total = row.total or 0
    auto_resolved = row.auto_resolved or 0
    manual_resolved = total - auto_resolved

    hourly_cost = metric.agent_hourly_cost if metric else DEFAULT_AGENT_HOURLY_COST
    manual_mins = metric.avg_manual_resolution_mins if metric else DEFAULT_AVG_MANUAL_MINS
    auto_mins = metric.avg_auto_resolution_mins if metric else DEFAULT_AVG_AUTO_MINS

    hours_saved = (auto_resolved * manual_mins) / 60
    cost_saved = hours_saved * hourly_cost
    capacity_freed_pct = round((auto_resolved / total * 100) if total > 0 else 0, 1)
    ai_cost_per_ticket = round((auto_mins / 60 * hourly_cost), 2)
    human_cost_per_ticket = round((manual_mins / 60 * hourly_cost), 2)

    return {
        "total_tickets": total,
        "auto_resolved_count": auto_resolved,
        "manual_resolved_count": manual_resolved,
        "auto_resolution_rate_percent": round((auto_resolved / total * 100) if total > 0 else 0, 1),
        "hours_saved": round(hours_saved, 1),
        "cost_saved_usd": round(cost_saved, 2),
        "agent_capacity_freed_percent": capacity_freed_pct,
        "ai_cost_per_ticket_usd": ai_cost_per_ticket,
        "human_cost_per_ticket_usd": human_cost_per_ticket,
        "roi_multiple": round(cost_saved / max((total - auto_resolved) * human_cost_per_ticket, 1), 2),
        "payback_period_months": round(100000 / max(cost_saved / max(total, 1) * 200, 1), 1),
        "settings": {
            "agent_hourly_cost": hourly_cost,
            "avg_manual_resolution_mins": manual_mins,
            "avg_auto_resolution_mins": auto_mins,
        },
    }


@router.get("/history")
async def roi_history(db: AsyncSession = Depends(get_db), agent=Depends(get_current_agent)):
    result = await db.execute(
        select(ROIMetric).order_by(ROIMetric.period_start.asc()).limit(12)
    )
    metrics = result.scalars().all()
    return [
        {
            "period_start": m.period_start,
            "period_end": m.period_end,
            "total_tickets": m.total_tickets,
            "auto_resolved_count": m.auto_resolved_count,
            "hours_saved": m.hours_saved,
            "cost_saved": m.cost_saved,
        }
        for m in metrics
    ]


@router.patch("/settings")
async def update_roi_settings(
    body: dict,
    db: AsyncSession = Depends(get_db),
    agent=Depends(get_current_agent),
):
    result = await db.execute(
        select(ROIMetric).order_by(ROIMetric.calculated_at.desc()).limit(1)
    )
    metric = result.scalar_one_or_none()
    if not metric:
        metric = ROIMetric()
        db.add(metric)

    if "agent_hourly_cost" in body:
        metric.agent_hourly_cost = float(body["agent_hourly_cost"])
    if "avg_manual_resolution_mins" in body:
        metric.avg_manual_resolution_mins = float(body["avg_manual_resolution_mins"])
    if "avg_auto_resolution_mins" in body:
        metric.avg_auto_resolution_mins = float(body["avg_auto_resolution_mins"])

    return {"success": True, "message": "ROI settings updated"}
